virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,340 @@
+
+    return (1.0e-6) * flux_to_energy * (1.0 - albedo_vis) * tau * shortwave_in
+
+
+def calculate_topofcanopy_radiation(
+    tau: DataArray,
+    shortwave_in: DataArray,
+    albedo_shortwave: Union[DataArray, float] = 0.17,
+) -> DataArray:
+    """Calculate top of canopy shortwave radiation, [J m-2].
+
+    Args:
+        tau: atmospheric transmissivity, unitless
+        shortwave_in: downward shortwave radiation, [J m-2]
+        albedo_shortwave: shortwave albedo, default = 0.17
+
+    Returns:
+        top of canopy radiation shortwave radiation, [J m-2]
+    """
+
+    return (1.0 - albedo_shortwave) * tau * shortwave_in
+
+
+def calculate_longwave_radiation(
+    temperature: DataArray,
+    emissivity: float = 0.95,
+    stefan_boltzmann_constant: float = RadiationConstants.stefan_boltzmann_constant,
+    celsius_to_kelvin: float = RadiationConstants.celsius_to_kelvin,
+) -> DataArray:
+    """Calculate longwave emission, [J m-2].
+
+    Longwave radiation is calculated following the Stefan-Boltzmann law as a function of
+    surface temperature (in Kelvin) to the fourth power.
+
+    Args:
+        temperature: canopy or soil temperature of n layers, [C]; the array size is
+            set to max number of layers (n_max) and filled with NaN where n < n_max
+        emissivity: canopy or soil emissivity, default=0.95
+        stefan_boltzmann_constant: Stefan-Boltzmann constant [W m-2 K-4]
+        celsius_to_kelvin: factor to convert temperature in Celsius to absolute
+            temperature in Kelvin
+
+    Returns:
+        longwave radiation from n individual canopy layers or soil, [J m-2]
+    """
+    # TODO add sanity check for temperature and in celsius?
+    # longwave emission canopy
+    return (
+        emissivity * stefan_boltzmann_constant * (celsius_to_kelvin + temperature) ** 4
+    )
+
+
+def calculate_netradiation_surface(
+    topofcanopy_radiation: DataArray,
+    canopy_absorption: DataArray,
+    longwave_canopy: DataArray,
+    longwave_soil: DataArray,
+) -> DataArray:
+    """Calculate net shortwave radiation at the surface, [J m-2].
+
+    The net shortwave radiation (RN) at the forest floor is calculated as
+    `topofcanopy_radiation` - `canopy_absorption` (summed over all canopy layers)
+    - `longwave_canopy` (summed over all canopy layers) - `longwave_soil`.
+
+    Canopy_absorption and longwave_canopy must include the dimension `canopy_layers` and
+    have the correct length (length of cell_id).
+
+    Args:
+        topofcanopy_radiation: top of canopy radiation shortwave radiation, [J m-2]
+        canopy_absorption: shortwave radiation absorbed by canopy layers, [J m-2]
+        longwave_canopy: longwave radiation from canopy layers, [J m-2]
+        longwave_soil: longwave radiation from soil layers, [J m-2]
+
+    Returns:
+        net shortwave radiation at the surface ( = forest floor), [J m-2]
+    """
+
+    # TODO raise error if canopy_layers don't have correct dimensions, not here?
```

Comment:
If `canopy_layers` doesn't have the correct dimensions, won't you get an error anyway?

Reply:
probably yes. I was just thinking of what could go wrong...

Reply:
You could add a test to check that the error *is* raised if a bad `canopy_layers` is passed in....